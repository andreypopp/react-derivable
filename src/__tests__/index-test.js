/**
 * @copyright 2016, Prometheus Research, LLC
 * @copyright 2017-present, Andrey Popp
 *
 * @flow
 */

import * as React from 'react';
import * as ReactDOMServer from 'react-dom/server';
import * as Derivable from 'derivable';
import {render as r, renderMap} from '../index';
import TestRenderer from 'react-test-renderer';

test('render', function() {
  const answer = Derivable.atom(42);

  function Component() {
    return <div>The answer is {r(answer)}</div>;
  }

  const renderer = TestRenderer.create(<Component />);
  expect(renderer.toJSON()).toMatchSnapshot();
  answer.set(0);
  expect(renderer.toJSON()).toMatchSnapshot();

  renderer.unmount();
});

test('renderMap', function() {
  const answer = Derivable.atom(42);

  function Component() {
    return <div>The answer is {renderMap(answer, answer => <span>{answer}</span>)}</div>;
  }

  const renderer = TestRenderer.create(<Component />);
  expect(renderer.toJSON()).toMatchSnapshot();
  answer.set(0);
  expect(renderer.toJSON()).toMatchSnapshot();

  renderer.unmount();
});

test('the number of derivation computations', function() {
  const effects = [];
  const answer = Derivable.atom(42);
  const answerPlusOne = answer.derive(answer => {
    effects.push('answerPlusOne');
    return answer + 1;
  });

  function Component() {
    return <div>The answer is {r(answerPlusOne)}</div>;
  }

  const renderer = TestRenderer.create(<Component />);
  expect(renderer.toJSON()).toMatchSnapshot();
  expect(effects).toEqual(['answerPlusOne']);
  answer.set(0);
  expect(renderer.toJSON()).toMatchSnapshot();
  expect(effects).toEqual(['answerPlusOne', 'answerPlusOne']);

  renderer.unmount();

  answer.set(1);
  expect(effects).toEqual(['answerPlusOne', 'answerPlusOne']);
});

test('rendering to string', function() {
  const answer = Derivable.atom(42);

  function Component() {
    return <div>The answer is {r(answer)}</div>;
  }

  expect(ReactDOMServer.renderToString(<Component />)).toMatchSnapshot();
});
