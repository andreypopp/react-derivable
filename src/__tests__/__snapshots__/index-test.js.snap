// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
<div>
  The answer is 
  42
</div>
`;

exports[`render 2`] = `
<div>
  The answer is 
  0
</div>
`;

exports[`renderMap 1`] = `
<div>
  The answer is 
  <span>
    42
  </span>
</div>
`;

exports[`renderMap 2`] = `
<div>
  The answer is 
  <span>
    0
  </span>
</div>
`;

exports[`rendering to string 1`] = `"<div data-reactroot=\\"\\">The answer is <!-- -->42</div>"`;

exports[`the number of derivation computations 1`] = `
<div>
  The answer is 
  43
</div>
`;

exports[`the number of derivation computations 2`] = `
<div>
  The answer is 
  1
</div>
`;
